src/kernels/jupyter/finder/universalRemoteKernelFinder.ts

```diff
@@ -0,0 +1,393 @@
+export class UniversalRemoteKernelFinder implements IRemoteKernelFinder, IContributedKernelFinderInfo, IDisposable {
+    /**
+     * List of ids of kernels that should be hidden from the kernel picker.
+     */
+    private readonly kernelIdsToHide = new Set<string>();
+    kind: string = 'remote';
+    id: string;
+    displayName: string;
+    private _cacheUpdateCancelTokenSource: CancellationTokenSource | undefined;
+    private cache: RemoteKernelConnectionMetadata[] = [];
+
+    private _onDidChangeKernels = new EventEmitter<void>();
+    onDidChangeKernels: Event<void> = this._onDidChangeKernels.event;
+
+    private _initializeResolve: () => void;
+    private _initializedPromise: Promise<void>;
+
+    private readonly disposables: IDisposable[] = [];
+
+    get initialized(): Promise<void> {
+        return this._initializedPromise;
+    }
+
+    private wasPythonInstalledWhenFetchingKernels = false;
+
+    constructor(
+        private jupyterSessionManagerFactory: IJupyterSessionManagerFactory,
+        private interpreterService: IInterpreterService,
+        private extensionChecker: IPythonExtensionChecker,
+        private readonly notebookProvider: INotebookProvider,
+        private readonly globalState: Memento,
+        private readonly env: IApplicationEnvironment,
+        private readonly cachedRemoteKernelValidator: IJupyterRemoteCachedKernelValidator,
+        kernelFinder: KernelFinder,
+        private readonly kernelProvider: IKernelProvider,
+        private readonly extensions: IExtensions,
+        private isWebExtension: boolean,
+        private readonly serverUri: IJupyterServerUriEntry
+    ) {
+        // Register with remote-serverId as our ID
+        this.id = `${this.kind}-${serverUri.serverId}`;
+
+        // Create a reasonable display name for this kernel finder
+        this.displayName = localize.DataScience.universalRemoteKernelFinderDisplayName().format(
+            serverUri.displayName || serverUri.uri
+        );
+
+        this._initializedPromise = new Promise<void>((resolve) => {
+            this._initializeResolve = resolve;
+        });
+
+        // When we register, add a disposable to clean ourselves up from the main kernel finder list
+        // Unlike the Local kernel finder universal remote kernel finders will be added on the fly
+        this.disposables.push(kernelFinder.registerKernelFinder(this));
+    }
+
+    dispose(): void | undefined {
+        disposeAllDisposables(this.disposables);
+    }
+
+    async activate(): Promise<void> {
+        // warm up the cache
+        this.loadCache().then(noop, noop);
+
+        // If we create a new kernel, we need to refresh if the kernel is remote (because
+        // we have live sessions possible)
+        // Note, this is a perf optimization for right now. We should not need
+        // to check for remote if the future when we support live sessions on local
+        this.kernelProvider.onDidStartKernel((k) => {
```

Comment:
Add this and listeners below to `this.disposables`?

Reply:
@rebornix Ahh, good catch. So the next two listeners below are added to disposables, they are just added via using the third function parameter. However this onDidStartKernel was missing that. It was copied over from remoteKernelFinder.ts so I also fixed it up there as well.